File: 02 09 25/findfirstindexinstring.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        if not needle:
            return 0

        # Build LPS array
        m = len(needle)
        lps = [0] * m
        length = 0
        i = 1
        while i < m:
            if needle[i] == needle[length]:
                length += 1
                lps[i] = length
                i += 1
            else:
                if length != 0:
                    length = lps[length - 1]
                else:
                    lps[i] = 0
                    i += 1

        # KMP search
        n = len(haystack)
        i = j = 0
        while i < n:
            if haystack[i] == needle[j]:
                i += 1
                j += 1
            else:
                if j != 0:
                    j = lps[j - 1]
                else:
                    i += 1
            if j == m:
                return i - m
        return -1

File: 02 09 25/findfirstindexinstring.rs
struct Solution;

impl Solution {
    pub fn str_str(haystack: String, needle: String) -> i32 {
        let n = haystack.len();
        let m = needle.len();
        if m == 0 { return 0; }
        if m > n { return -1; }

        for i in 0..=n - m {
            if &haystack[i..i + m] == needle {
                return i as i32;
            }
        }
        -1
    }

    pub fn str_str_built_in(haystack: String, needle: String) -> i32 {
        haystack.find(&needle).map(|i| i as i32).unwrap_or(-1)
    }

fn main() {
    println!("{}", Solution::str_str("hello".to_string(), "ll".to_string())); // 2
    println!("{}", Solution::str_str("aaaaa".to_string(), "bba".to_string())); // -1
}
